Add ``refresh_cache`` option to ``EarthLocation`` methods

tbowers7 opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

Add the new option ``refresh_cache`` option to ``EarthLocation.of_site()`` and ``EarthLocation.get_site_names()`` which is then passed along to the ``force_download`` option of ``EarthLocation._get_site_registry()``.  This last method already had the ``force_download`` option; this PR simply adds a user-facing option to the public methods.


<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pllim commented:
Why not use `cache` (bool) to be consistent with `astropy.utils.data.download_file` public API?

tbowers7 commented:
> Why not use `cache` (bool) to be consistent with `astropy.utils.data.download_file` public API?

My impression that `astropy.utils.data.download_file` is aimed at developers whereas the methods in this PR are aimed at the end user.  Furthermore, the `cache` (bool) from the utils package is asking whether to cache downloaded file or not, and its meaning is less clear to the user than `force_download`.

If there is a compelling reason to use `cache` (bool), and its meaning can be made plain (along with minimal additional logic to be added to these methods) or some other keyword parameter, I am happy to do so.

eerovaher commented:
The problem this pull request is trying to address is that once the observatory site registry is downloaded `astropy` will never try to update it on its own. If a user wants a newer version of the registry they either have to delete the registry cache or use the `force_download` option of the _private_ `EarthLocation._get_site_registry()` method. The solution this pull request proposes is to make the `force_download` option accessible from public `EarthLocation` methods. That would indeed be an improvement, but I think at minimum the users should be told about this in the error message they receive if a name they requested does not match or is not similar to anything in the registry. But it is also worth considering if the local site registry cache should expire periodically (see #1162).

I am also not sure if the name `force_download` is the best. It is the name a private method already uses, but that does not mean the public API should be obligated to use the same name. Something like `refresh_cache` might be a better name for a public method argument.

tbowers7 commented:
Thank you @eerovaher for your comments.

I have updated the parameter to the ``EarthLocation`` methods from ``force_download`` to ``refresh_cache`` and added the requested error verbiage to alert users to the ability to download a fresh version of the site registry.

I will leave the question of periodical expiration of the local site registry cache to others.

eerovaher commented on astropy/coordinates/earth.py:
```diff
@@ -342,10 +342,19 @@ def of_site(cls, site_name):
         site_name : str
             Name of the observatory (case-insensitive).
+        refresh_cache : bool, optional
+            If True, force replacement of the cached registry with a
+            downloaded version.  (Default: False)
```
```suggestion
            If `True`, force replacement of the cached registry with a
            newly downloaded version.  (Default: `False`)
            
            .. versionadded:: 5.3
```

tbowers7 replied:
Changed.

eerovaher reviewed (changes requested):
Exposing the option to refresh the cache in the public API could be very useful, but I don't think this pull request is ready for merging quite yet. I've made a few comments regarding `EarthLocation.of_site()`, but analogous comments also apply to `EarthLocation.get_site_names()`. Furthermore, grepping revealed the following warning message: https://github.com/astropy/astropy/blob/0ddc345c58ae04c5383713d54c01dd39aa7d5a72/astropy/coordinates/earth.py#L557-L562
If the public API provides the `refresh_cache` option then the warning message should not suggest calling a private method.

The documentation at https://docs.astropy.org/en/latest/coordinates/index.html#convenience-methods should be updated.

There should be a unit test that would check if the `refresh_cache` value is passed on to `EarthLocation._get_site_registry()` with the correct keyword. This might be easiest to achieve by monkeypatching `_get_site_registry()`.

tbowers7 commented:
Requested changes to the documentation and warning messages have been made.

> There should be a unit test that would check if the ``refresh_cache value`` is passed on to ``EarthLocation._get_site_registry()`` with the correct keyword. This might be easiest to achieve by monkeypatching ``_get_site_registry()``.

I'm afraid I don't know how to do this, and will require assistance.

tbowers7 commented on astropy/coordinates/earth.py:
```diff
@@ -319,7 +319,7 @@ def from_geodetic(cls, lon, lat, height=0.0, ellipsoid=None):
 
     @classmethod
-    def of_site(cls, site_name):
+    def of_site(cls, site_name, refresh_cache=False):
```
Changed to make ``refresh-cache`` keyword-only.

tbowers7 commented on astropy/coordinates/earth.py:
```diff
@@ -342,10 +342,19 @@ def of_site(cls, site_name):
+            version of the site registry even if the version in the
+            `astropy-data repository <https://github.com/astropy/astropy-data>`_
+            has been updated.  To refresh the local cache to use an updated
+            site registry, use the ``refresh_cache=True`` option.
```
Paragraph has been rewritten for clarity.

eerovaher commented:
I've implemented the unit tests and I've updated the documentation. @tbowers7, if you agree with how I've updated the pull request then I ask you to squash the current commits down to one and rebase on the `main` branch. Furthermore, because I believe users would benefit from being exposed to the new functionality I also ask that you add a second commit with an entry to `docs/whatsnew/5.3.rst`.

You might find it useful to know that calling `git show --name-status` (for one commit) or `git log --name-status` will show what files have been edited, so there's no need to write down a list in the commit message.

pllim commented on astropy/coordinates/earth.py:
```diff
@@ -555,12 +571,11 @@ def _get_site_registry(cls, force_download=False, force_builtin=False):
+                        "retry the download, use the option `refresh_cache=True`."
                     )
-                    warn(AstropyUserWarning(msg.format(cls.__name__)))
+                    warn(AstropyUserWarning(msg))
```
While we're at this, isn't this a more proper usage?
```suggestion
                    warn(msg, AstropyUserWarning)
```

eerovaher replied:
Yes indeed. Furthermore, this warning isn't being tested. @tbowers7, are you interested in applying the suggestion by @pllim and adding a unit test for this or do you think this is too far beyond the scope of your original idea?

pllim replied:
To be sure, you can generate the coverage report locally when you test this branch and enable `--remote-data` pytest flag.

eerovaher replied:
This warning is emitted if accessing the remote server fails, so there could and should be a test for this that doesn't require remote data.

tbowers7 replied:
It sounds like a unit test for this warning should be created, but I would suggest this be left as an [issue](https://github.com/astropy/astropy/issues) separate from this PR.  I would, however, be willing to entertain a friendly commit that implements this unit test from someone more familiar with the intricacies of AstroPy's testing structures.

eerovaher replied:
I don't think this warning should hold back this pull request, I'd rather have a follow-up pull request add the missing test.

pllim commented on astropy/coordinates/earth.py:
```diff
@@ -555,12 +571,11 @@ def _get_site_registry(cls, force_download=False, force_builtin=False):
-                        "{0}._get_site_registry(force_download=True)"
+                        "Could not access the main site list. Falling back on the "
+                        "built-in version, which is rather limited. If you want to "
+                        "retry the download, use the option `refresh_cache=True`."
```
I don't think single backticks do anything here since this warning would be issued in plain text?
```suggestion
                        "retry the download, use the option refresh_cache=True."
```

pllim replied:
Single quote might be easier on the eyes in plain text. At least that is my personal opinion. I am not fonts expert.

pllim commented on astropy/coordinates/errors.py:
```diff
@@ -193,6 +193,8 @@ class UnknownSiteException(KeyError):
         message = (
             f"Site '{site}' not in database. Use {attribute} to see available sites."
+            f" If '{site}' exists in the online astropy-data repository, use the"
+            " `refresh_cache` option to download the latest version."
```
```suggestion
            " refresh_cache option to download the latest version."
```

pllim commented on astropy/coordinates/tests/test_sites.py:
```diff
@@ -86,6 +86,38 @@ def test_EarthLocation_basic():
+    "class_method,args",
+    [(EarthLocation.get_site_names, []), (EarthLocation.of_site, ["greenwich"])],
+)
+def test_Earthlocation_refresh_cache_is_mandatory_kwarg(args, class_method):
```
It is confusing to have the parametrize order to be opposite of actual args order.
```suggestion
def test_Earthlocation_refresh_cache_is_mandatory_kwarg(class_method, args):
```

eerovaher replied:
I ordered the arguments alphabetically, but the alphabetical ordering is not important here so it might not have been helpful.

tbowers7 replied:
I will leave issues with the unit test to @eerovaher to resolve.

pllim commented on astropy/coordinates/tests/test_sites.py:
```diff
@@ -86,6 +86,38 @@ def test_EarthLocation_basic():
+    [(EarthLocation.get_site_names, []), (EarthLocation.of_site, ["greenwich"])],
+)
+@pytest.mark.parametrize("refresh_cache", [False, True])
+def test_Earthlocation_refresh_cache(args, class_method, monkeypatch, refresh_cache):
```
```suggestion
def test_Earthlocation_refresh_cache(class_method, args, monkeypatch, refresh_cache):
```

tbowers7 replied:
I will leave issues with the unit test to @eerovaher to resolve.

pllim commented on docs/coordinates/index.rst:
```diff
@@ -340,6 +340,19 @@ To see the list of site names available, use
     >>> EarthLocation.get_site_names()  # doctest: +REMOTE_DATA
     ['ALMA', 'AO', 'ARCA', ...]
 
+Both :func:`~astropy.coordinates.EarthLocation.of_site` and
+:func:`~astropy.coordinates.EarthLocation.get_site_names`,
```
```suggestion
Both :meth:`~astropy.coordinates.EarthLocation.of_site` and
:meth:`~astropy.coordinates.EarthLocation.get_site_names`,
```

tbowers7 replied:
Moved the references to ``:func:`` to ``:meth:`` in this section of the documentation.

eerovaher commented on docs/whatsnew/5.3.rst:
```diff
@@ -181,6 +182,22 @@ we can take the sum of ND masked quantities along the ``1`` axis like so::
 
+.. _whatsnew-5.3-coordinates-refresh-site-registry:
+
+Refresh cached site registry for ``coordinates.EarthLocation`` methods
+======================================================================
```
Differently from the change log entries a What's New entry should contain more links rather than less:
```suggestion
Refresh cached site registry for |EarthLocation| methods
========================================================
```

eerovaher commented on docs/whatsnew/5.3.rst:
```diff
@@ -181,6 +182,22 @@ we can take the sum of ND masked quantities along the ``1`` axis like so::
+Refresh cached site registry for ``coordinates.EarthLocation`` methods
+======================================================================
+
+The ``EarthLocation`` convenience methods
```
```suggestion
The |EarthLocation| convenience methods
```

eerovaher reviewed (commented):
@tbowers7, there are only a few minor formatting nitpicks and the missing unit test that can be considered to be beyond the scope of your idea.

pllim reviewed (approved):
LGTM. Thanks!

pllim commented:
@eerovaher , I agree and it seems like the author deferred to you on how to wrap this up, so since you are happy, I am merging now. Thanks, all!

pllim commented:
Please do open the follow up issue as you are more familiar with the details, and link it back to this PR. 🙇‍♀️
